### PR TITLE
Kernel/IPC: Implement StaticBuffer translation for HLE services that use the HLERequestContext architecture.

### PR DESCRIPTION
**Pull request by Subv**
The real kernel requires services to set up their static buffer targets ahead of time. This implementation does not require that and will simply create the storage for the buffers as they are processed in the incoming IPC request.

Static buffers are kept in an unordered_map keyed by their buffer id, and are written into the already-setup area of the request thread when responding an IPC request.

This fixes a regression (crash) introduced in #2992.

This PR introduces more warnings due to the [[deprecated]] attribute being added to void PushStaticBuffer(VAddr buffer_vaddr, size_t size, u8 buffer_id); and VAddr PopStaticBuffer(size_t* data_size);

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3077)

### CODE REVIEW

File: src/core/hle/ipc.h [hunk 1]
```diff
@@ -40,13 +40,19 @@ static const int kStaticBuffersOffset = 0x100;
...
-}
+} // namespace Kernel
 
 namespace IPC {
 
 /// Size of the command buffer area, in 32-bit words.
 constexpr size_t COMMAND_BUFFER_LENGTH = 0x100 / sizeof(u32);
 
+// Offset from the command buffer area of the static buffers, in 32-bit words.
+constexpr size_t STATIC_BUFFERS_OFFSET = Kernel::kStaticBuffersOffset / sizeof(u32);
```
Comment (wwylele) [inline src/core/hle/ipc.h]: As `Kernel::kStaticBuffersOffset = 0x100` (the static buffer descriptor area follows the end of command buffer), it is probably better to unify these two constants `COMMAND_BUFFER_LENGTH` and `STATIC_BUFFERS_OFFSET` to make it clear.

File: src/core/hle/kernel/hle_ipc.cpp [hunk 2]
```diff
@@ -124,6 +147,25 @@ ResultCode HLERequestContext::WriteToOutgoingCommandBuffer(u32_le* dst_cmdbuf, P
...
+            IPC::StaticBufferDescInfo bufferInfo{descriptor};
+
+            const auto& data = GetStaticBuffer(bufferInfo.buffer_id);
+
+            // Grab the address that the target thread set up to receive the response static buffer
+            // and write our data there.
+            IPC::StaticBufferDescInfo target_descriptor{
+                dst_cmdbuf[IPC::STATIC_BUFFERS_OFFSET * bufferInfo.buffer_id]};
+            VAddr target_address =
+                dst_cmdbuf[IPC::STATIC_BUFFERS_OFFSET * bufferInfo.buffer_id + 1];
```
Comment (wwylele) [inline src/core/hle/kernel/hle_ipc.cpp]: > `IPC::STATIC_BUFFERS_OFFSET * bufferInfo.buffer_id`

I think you want to say `IPC::STATIC_BUFFERS_OFFSET + 2 * bufferInfo.buffer_id`

**Comment by wwylele** [comment]
Re test: could you also append the `translates mixed params` section? Also please fix clang-format problem

**Comment by yuriks** [comment]
Reviewed 8 of 10 files at r1, 2 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[src/core/hle/ipc.h, line 51 at r1](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyCIoG315TeztAySECr-r1-51:-KyOvDZ0kjyZFVsqpmOk:b-hhwtr3) ([raw file](https://github.com/citra-emu/citra/blob/972a24cf8037fdcc8a20ee2e361f226e9565211d/src/core/hle/ipc.h#L51)):*
<details><summary><i>Previously, wwylele (Weiyi Wang) wrote…</i></summary><blockquote>

As `Kernel::kStaticBuffersOffset = 0x100` (the static buffer descriptor area follows the end of command buffer), it is probably better to unify these two constants `COMMAND_BUFFER_LENGTH` and `STATIC_BUFFERS_OFFSET` to make it clear.
</blockquote></details>

+1, even if it it's just aliasing one name to the other.

---

*[src/core/hle/ipc_helpers.h, line 122 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOpr3g_7oKWnQ5uK2a:-KyOpr3g_7oKWnQ5uK2b:bv6yrye) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/ipc_helpers.h#L122)):*
> ```objc
> 
>     [[deprecated]] void PushStaticBuffer(VAddr buffer_vaddr, size_t size, u8 buffer_id);
> ```

nit: Keep these two lines grouped together? (Since they're 2 overloads of the same thing...)

---

*[src/core/hle/ipc_helpers.h, line 206 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOpzKWsgr2iFmn60-o:-KyOpzKWsgr2iFmn60-p:b-u45ry8) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/ipc_helpers.h#L206)):*
> ```objc
>     Push(StaticBufferDesc(buffer.size(), buffer_id));
>     // This address will be replaced by the correct static buffer address during IPC translation.
>     Push<VAddr>(0);
> ```

Since we ignore it on our end anyway, I'd try to use some obvious poison value here.

---

*[src/core/hle/ipc_helpers.h, line 486 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOqV_vXGhwOcSAHmWk:-KyOqV_vXGhwOcSAHmWl:b-p3t32q) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/ipc_helpers.h#L486)):*
> ```objc
>     Pop<VAddr>();
> 
>     StaticBufferDescInfo bufferInfo{sbuffer_descriptor};
> ```

`buffer_info`

---

*[src/core/hle/kernel/hle_ipc.cpp, line 160 at r1](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyCIoG315TeztAySECt-r1-160:-KyOr5sHrHj2fkdjs_Sy:b-c6imhr) ([raw file](https://github.com/citra-emu/citra/blob/972a24cf8037fdcc8a20ee2e361f226e9565211d/src/core/hle/kernel/hle_ipc.cpp#L160)):*
<details><summary><i>Previously, wwylele (Weiyi Wang) wrote…</i></summary><blockquote>

> `IPC::STATIC_BUFFERS_OFFSET * bufferInfo.buffer_id`

I think you want to say `IPC::STATIC_BUFFERS_OFFSET + 2 * bufferInfo.buffer_id`
</blockquote></details>

+1. Also, use a variable for `IPC::STATIC_BUFFERS_OFFSET + 2 * bufferInfo.buffer_id` instead of repeating it twice.

---

*[src/core/hle/kernel/hle_ipc.cpp, line 54 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOnlypboAbcAo3gv-H:-KyOnlypboAbcAo3gv-I:by4w3wm) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/kernel/hle_ipc.cpp#L54)):*
> ```cpp
> }
> 
> void HLERequestContext::AddStaticBuffer(u8 buffer_id, const std::vector<u8>& data) {
> ```

You should take this argument by-value so that a vector can be moved into it.

---

*[src/core/hle/kernel/hle_ipc.cpp, line 55 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOqcKIvGSM8ZBnvTtn:-KyOqcKIvGSM8ZBnvTto:b-p03gxd) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/kernel/hle_ipc.cpp#L55)):*
> ```cpp
> 
> void HLERequestContext::AddStaticBuffer(u8 buffer_id, const std::vector<u8>& data) {
>     static_buffers[buffer_id] = data;
> ```

There should be an assert against setting the same buffer_id more than once.

---

*[src/core/hle/kernel/hle_ipc.cpp, line 100 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOrWZTnHqJ01wv8kit:-KyOrWZTnHqJ01wv8kiu:b-p3t32q) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/kernel/hle_ipc.cpp#L100)):*
> ```cpp
>         case IPC::DescriptorType::StaticBuffer: {
>             VAddr source_address = src_cmdbuf[i];
>             IPC::StaticBufferDescInfo bufferInfo{descriptor};
> ```

`buffer_info`

---

*[src/core/hle/kernel/hle_ipc.cpp, line 107 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOqu-kqmNol7cockkQ:-KyOqu-kqmNol7cockkR:b-5xupgh) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/kernel/hle_ipc.cpp#L107)):*
> ```cpp
> 
>             static_buffers[bufferInfo.buffer_id] = std::move(data);
>             cmd_buf[i++] = source_address;
> ```

Add an assert like in the other case to make sure that this read won't overflow the command size.

---

*[src/core/hle/kernel/hle_ipc.cpp, line 163 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOrRFAlICqh6Sq1lBB:-KyOrRFAlICqh6Sq1lBC:bjjiax3) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/kernel/hle_ipc.cpp#L163)):*
> ```cpp
>                 dst_cmdbuf[IPC::COMMAND_BUFFER_LENGTH + 2 * bufferInfo.buffer_id + 1];
> 
>             ASSERT_MSG(target_descriptor.size >= data.size(), "Static buffer data is too big");
> ```

Wouldn't these return translation errors?

---

*[src/core/hle/kernel/hle_ipc.cpp, line 167 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOrFeMazw6g7Le8WF2:-KyOrFeMazw6g7Le8WF3:bir6iiq) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/kernel/hle_ipc.cpp#L167)):*
> ```cpp
>             Memory::WriteBlock(dst_process, target_address, data.data(), data.size());
> 
>             dst_cmdbuf[i++] = target_address;
> ```

Add an assert to ensure this won't overflow the command size.

---

*[src/core/hle/kernel/hle_ipc.h, line 148 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOoJhdznQX58BsClDP:-KyOoJhdznQX58BsClDQ:bven5hm) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/kernel/hle_ipc.h#L148)):*
> ```objc
>     boost::container::small_vector<SharedPtr<Object>, 8> request_handles;
>     // The static buffers will be created when the IPC request is translated.
>     std::unordered_map<u8, std::vector<u8>> static_buffers;
> ```

There can only bet 16 static buffers. Just make this an array of 16 vectors and directly index into it (with appropriate bounds checks). (Empty vectors don't allocate.)

If you want to save space on the struct for the common case of no static buffers (since the array would be 16*24=384 bytes) you can use a vector of vectors that gets initialized on first use. (Just stick a static_buffers.resize(16); before it's used.)

---

*[src/core/hle/service/ac/ac.cpp, line 28 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOslZKzA4w-iRtqzvM:-KyOslZKzA4w-iRtqzvN:b-83gdrd) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/service/ac/ac.cpp#L28)):*
> ```cpp
>     IPC::RequestBuilder rb = rp.MakeBuilder(1, 2);
>     rb.Push(RESULT_SUCCESS);
>     rb.PushStaticBuffer(buffer, 0);
> ```

(After you change the parameter type.) std::move

---

*[src/core/hle/service/ac/ac.cpp, line 126 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOt5r8I27kvg7fEyDA:-KyOt5r8I27kvg7fEyDB:b-83gdrd) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/service/ac/ac.cpp#L126)):*
> ```cpp
>     IPC::RequestBuilder rb = rp.MakeBuilder(1, 2);
>     rb.Push(RESULT_SUCCESS);
>     rb.PushStaticBuffer(ac_config, 0);
> ```

(After you change the parameter type.) std::move

---

*[src/tests/core/hle/kernel/hle_ipc.cpp, line 136 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOtmqyXrVpuUjKWjm5:-KyOtmqyXrVpuUjKWjm6:b-jd4pmo) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/tests/core/hle/kernel/hle_ipc.cpp#L136)):*
> ```cpp
>         CHECK(context.GetStaticBuffer(0) == *buffer);
> 
>         REQUIRE(process->vm_manager.UnmapRange(target_address, buffer->size()) == RESULT_SUCCESS);
> ```

Is this cleanup important? You can just let the process be destroyed.

---

*[src/tests/core/hle/kernel/hle_ipc.cpp, line 170 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOuL2kUe7f646GP2QB:-KyOuL2kUe7f646GP2QC:b10kmjp) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/tests/core/hle/kernel/hle_ipc.cpp#L170)):*
> ```cpp
>         CHECK(context.GetStaticBuffer(0) == *buffer);
> 
>         REQUIRE(process->vm_manager.UnmapRange(target_address, buffer->size()) == RESULT_SUCCESS);
> ```

Same here, seems like this could just be omitted.

---

*[src/tests/core/hle/kernel/hle_ipc.cpp, line 276 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOuzDZp-83EhmcATN9:-KyOuzDZp-83EhmcATNA:b-o2zdkl) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/tests/core/hle/kernel/hle_ipc.cpp#L276)):*
> ```cpp
>         CHECK(*output_buffer == input_buffer);
>         REQUIRE(process->vm_manager.UnmapRange(target_address, output_buffer->size()) ==
>                 RESULT_SUCCESS);
> ```

Again, same thing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/3077)*

**Comment by Dragios** [comment]
Also fix crash after click "Touch to continue" in Zen Pinball 3D

This was the log generated:
`Debug <Critical> core/hle/kernel/hle_ipc.cpp:operator():128: Assertion Failed!
Unsupported handle translation: 0x00800002`

**Comment by Subv** [comment]
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[src/core/hle/ipc.h, line 51 at r1](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyCIoG315TeztAySECr-r1-51:-KyksBig1iLHrQmtIN0T:b-f5h4c4) ([raw file](https://github.com/citra-emu/citra/blob/972a24cf8037fdcc8a20ee2e361f226e9565211d/src/core/hle/ipc.h#L51)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

+1, even if it it's just aliasing one name to the other.
</blockquote></details>

That constant does not exist anymore

---

*[src/core/hle/ipc_helpers.h, line 122 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOpr3g_7oKWnQ5uK2a:-KyksrPP_M9uZaLs9VK7:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/ipc_helpers.h#L122)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

nit: Keep these two lines grouped together? (Since they're 2 overloads of the same thing...)
</blockquote></details>

Done.

---

*[src/core/hle/ipc_helpers.h, line 206 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOpzKWsgr2iFmn60-o:-Kykt7_beGzHeO_yxjzB:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/ipc_helpers.h#L206)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Since we ignore it on our end anyway, I'd try to use some obvious poison value here.
</blockquote></details>

Done.

---

*[src/core/hle/ipc_helpers.h, line 486 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOqV_vXGhwOcSAHmWk:-KyktZeKdIG7J3lpPEnV:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/ipc_helpers.h#L486)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

`buffer_info`
</blockquote></details>

Done.

---

*[src/core/hle/kernel/hle_ipc.cpp, line 160 at r1](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyCIoG315TeztAySECt-r1-160:-KykuvNMnbvsVZP64AnL:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/972a24cf8037fdcc8a20ee2e361f226e9565211d/src/core/hle/kernel/hle_ipc.cpp#L160)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

+1. Also, use a variable for `IPC::STATIC_BUFFERS_OFFSET + 2 * bufferInfo.buffer_id` instead of repeating it twice.
</blockquote></details>

Done.

---

*[src/core/hle/kernel/hle_ipc.cpp, line 54 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOnlypboAbcAo3gv-H:-KykteA1QUA7aBZfs5b8:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/kernel/hle_ipc.cpp#L54)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

You should take this argument by-value so that a vector can be moved into it.
</blockquote></details>

Done.

---

*[src/core/hle/kernel/hle_ipc.cpp, line 55 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOqcKIvGSM8ZBnvTtn:-Kyku4UtBATTDQeDdz6Z:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/kernel/hle_ipc.cpp#L55)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

There should be an assert against setting the same buffer_id more than once.
</blockquote></details>

Done.

---

*[src/core/hle/kernel/hle_ipc.cpp, line 100 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOrWZTnHqJ01wv8kit:-KykubprZC2TdgFYY6Xa:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/kernel/hle_ipc.cpp#L100)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

`buffer_info`
</blockquote></details>

Done.

---

*[src/core/hle/kernel/hle_ipc.cpp, line 107 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOqu-kqmNol7cockkQ:-KykuskYceMZs8nEH_dn:b-lbut5x) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/kernel/hle_ipc.cpp#L107)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Add an assert like in the other case to make sure that this read won't overflow the command size.
</blockquote></details>

What do you mean?

---

*[src/core/hle/kernel/hle_ipc.cpp, line 163 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOrRFAlICqh6Sq1lBB:-Kykux8mHocxKWgVhLQp:b-wlbsxt) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/kernel/hle_ipc.cpp#L163)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Wouldn't these return translation errors?
</blockquote></details>

Nope, the real kernel actually just panics when this happens

---

*[src/core/hle/kernel/hle_ipc.cpp, line 167 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOrFeMazw6g7Le8WF2:-KykvDlX9UPzTvVl22J6:bfd6trn) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/kernel/hle_ipc.cpp#L167)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Add an assert to ensure this won't overflow the command size.
</blockquote></details>

I don't understand what you mean by this

---

*[src/core/hle/kernel/hle_ipc.h, line 148 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOoJhdznQX58BsClDP:-Kykw6kc3pLCMEUsWBab:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/kernel/hle_ipc.h#L148)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

There can only bet 16 static buffers. Just make this an array of 16 vectors and directly index into it (with appropriate bounds checks). (Empty vectors don't allocate.)

If you want to save space on the struct for the common case of no static buffers (since the array would be 16*24=384 bytes) you can use a vector of vectors that gets initialized on first use. (Just stick a static_buffers.resize(16); before it's used.)
</blockquote></details>

Done.

---

*[src/core/hle/service/ac/ac.cpp, line 28 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOslZKzA4w-iRtqzvM:-KykwPVzPpbnsLK-QjX8:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/service/ac/ac.cpp#L28)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

(After you change the parameter type.) std::move
</blockquote></details>

Done.

---

*[src/core/hle/service/ac/ac.cpp, line 126 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOt5r8I27kvg7fEyDA:-KykwQ4KZy7TdK2wkldS:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/service/ac/ac.cpp#L126)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

(After you change the parameter type.) std::move
</blockquote></details>

Done.

---

*[src/tests/core/hle/kernel/hle_ipc.cpp, line 136 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOtmqyXrVpuUjKWjm5:-KykwYTbu1Nf2sVmXw16:b-x4yy8h) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/tests/core/hle/kernel/hle_ipc.cpp#L136)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Is this cleanup important? You can just let the process be destroyed.
</blockquote></details>

IMO it's better to keep each test case isolated and retain complete control of what goes on inside each SECTION.
Also since the Process is kept alive for all the cases in this block the allocation could interfere with other tests that also need to allocate memory

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/3077)*

**Comment by yuriks** [comment]
Late to the party~~~

---

Reviewed 4 of 5 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[src/core/hle/kernel/hle_ipc.cpp, line 107 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOqu-kqmNol7cockkQ:-L0-18p4AjRosEA1BEd1:bq49nr8) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/kernel/hle_ipc.cpp#L107)):*
<details><summary><i>Previously, Subv (Sebastian Valle) wrote…</i></summary><blockquote>

What do you mean?
</blockquote></details>

For example, Copy/MoveHandle has an assert that will check that i + num_handles won't overflow the command size before reading them. Likewise here there should be something similar that check that cmd_buff[i] is a valid position in the buffer.

---

*[src/core/hle/kernel/hle_ipc.cpp, line 167 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOrFeMazw6g7Le8WF2:-L0-1cgXk3R4x9HmnlPI:b4k8wb) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/core/hle/kernel/hle_ipc.cpp#L167)):*
<details><summary><i>Previously, Subv (Sebastian Valle) wrote…</i></summary><blockquote>

I don't understand what you mean by this
</blockquote></details>

Same as the other comment I explained above.

---

*[src/tests/core/hle/kernel/hle_ipc.cpp, line 136 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3077#-KyOtmqyXrVpuUjKWjm5:-L0-2JqutDXZC7UfOeQb:b9ift7u) ([raw file](https://github.com/citra-emu/citra/blob/3840d40e0a6cd29d7723183ba1c51a15a274914b/src/tests/core/hle/kernel/hle_ipc.cpp#L136)):*
<details><summary><i>Previously, Subv (Sebastian Valle) wrote…</i></summary><blockquote>

IMO it's better to keep each test case isolated and retain complete control of what goes on inside each SECTION.
Also since the Process is kept alive for all the cases in this block the allocation could interfere with other tests that also need to allocate memory
</blockquote></details>

Catch doesn't run all the test-cases in sequence like that. Each SECTION will have its own full full execution of the setup and teardown code (but only one section will be executed at a time). So there will be no leftover objects to interfere with other test cases even if you don't clean them up, because they'll fall out of scope.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/3077#-:-L0-2fR5S_HcaiMvVMTj:beo62s7)*

**Pull request merged**